rocksdb/rocksdb.nim

```diff
@@ -0,0 +1,251 @@
+  readopts,
+  writeopts,
+  cfdescriptor
+
+type
+  RocksDBResult*[T] = Result[T, string]
+
+  RocksDbPtr = ptr rocksdb_t
+
+  RocksDbRef* = ref object of RootObj
+    cPtr: RocksDbPtr
+    path: string
+    dbOpts: DbOptionsRef
+    readOpts: ReadOptionsRef
+    cfTable: ColFamilyTableRef
+
+  RocksDbReadOnlyRef* = ref object of RocksDbRef
+
+  RocksDbReadWriteRef* = ref object of RocksDbRef
+    writeOpts: WriteOptionsRef
+
+  DataProc* = proc(val: openArray[byte]) {.gcsafe, raises: [].}
+
+proc openRocksDb*(
+    path: string,
+    dbOpts = defaultDbOptions(),
+    readOpts = defaultReadOptions(),
+    writeOpts = defaultWriteOptions(),
+    columnFamilies = @[defaultColFamilyDescriptor()]): RocksDBResult[RocksDbReadWriteRef] =
+
+  if columnFamilies.len == 0:
+    return err("rocksdb: no column families")
+
+  var
+    cfNames = columnFamilies.mapIt(it.name().cstring)
+    cfOpts = columnFamilies.mapIt(it.options.cPtr)
+    columnFamilyHandles = newSeq[ColFamilyHandlePtr](columnFamilies.len)
+    errors: cstring
+
+  let rocksDbPtr = rocksdb_open_column_families(
+        dbOpts.cPtr,
+        path.cstring,
+        cfNames.len().cint,
+        cast[cstringArray](cfNames[0].addr),
+        cfOpts[0].addr,
+        columnFamilyHandles[0].addr,
+        cast[cstringArray](errors.addr))
+  bailOnErrors(errors)
+
+  var cfTable = newColFamilyTable()
+  for i, cf in columnFamilies:
+    cfTable.put(cf.name(), columnFamilyHandles[i])
+
+  let db = RocksDbReadWriteRef(
+      cPtr: rocksDbPtr,
+      path: path,
+      dbOpts: dbOpts,
+      readOpts: readOpts,
+      writeOpts: writeOpts,
+      cfTable: cfTable)
+  ok(db)
+
+proc openRocksDbReadOnly*(
+    path: string,
+    dbOpts = defaultDbOptions(),
+    readOpts = defaultReadOptions(),
+    columnFamilies = @[defaultColFamilyDescriptor()],
+    errorIfWalFileExists = false): RocksDBResult[RocksDbReadOnlyRef] =
+
+  if columnFamilies.len == 0:
+    return err("rocksdb: no column families")
+
+  var
+    cfNames = columnFamilies.mapIt(it.name().cstring)
+    cfOpts = columnFamilies.mapIt(it.options.cPtr)
+    columnFamilyHandles = newSeq[ColFamilyHandlePtr](columnFamilies.len)
+    errors: cstring
+
+  let rocksDbPtr = rocksdb_open_for_read_only_column_families(
+        dbOpts.cPtr,
+        path.cstring,
+        cfNames.len().cint,
+        cast[cstringArray](cfNames[0].addr),
+        cfOpts[0].addr,
+        columnFamilyHandles[0].addr,
+        errorIfWalFileExists.uint8,
+        cast[cstringArray](errors.addr))
+  bailOnErrors(errors)
+
+  var cfTable = newColFamilyTable()
+  for i, cf in columnFamilies:
+    cfTable.put(cf.name(), columnFamilyHandles[i])
+
+  let db = RocksDbReadOnlyRef(
+      cPtr: rocksDbPtr,
+      path: path,
+      dbOpts: dbOpts,
+      readOpts: readOpts,
+      cfTable: cfTable)
+  ok(db)
+
+template isClosed*(db: RocksDbRef): bool =
+  db.cPtr.isNil()
+
+proc cPtr*(db: RocksDbRef): RocksDbPtr =
+  doAssert not db.isClosed()
+  db.cPtr
+
+proc get*(
```

Comment:
This is just my idea. I think when `put`, `get`, `delete`, etc where `columnFamily` is expected, we can create a `ColumnFamily` object and use the overloaded version of `put`, `get`, `delete` where user don't have to write the `columnFamily`.
I think it will be more ergonomic.

Reply:
Yes good idea, perhaps that can be a secondary way of using a column family. One of my goals when re-writing the library was to follow the style/naming/patterns of the official Java RocksDB library and you can see they supply the column family handle in the get/put/delete methods etc. See here: https://javadoc.io/doc/org.rocksdb/rocksdbjni/latest/org/rocksdb/RocksDB.html. I opted to use a string instead of exposing the handle in the API for simplicity.

Reply:
For this I've opted to add a way to set the default column family on the RocksDbRef (in the next PR). It will provide a way to interact with a specific column family without having to pass it into every operation.